src/photon/photon_algorithm.c

```diff
@@ -115,21 +131,79 @@ void provide_scheduler_info(local_scheduler_state *state,
 }
 
 /**
+ * Record a queued task's missing object dependency.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param task_entry The task's queue entry.
+ * @param obj_id The ID of the object that the task is dependent on.
+ * @returns Void.
+ */
+void add_task_dependency(local_scheduler_state *state,
+                         scheduling_algorithm_state *algorithm_state,
+                         task_queue_entry *task_entry,
+                         object_id obj_id) {
+  object_dependency *entry;
+  HASH_FIND(dependency_handle, algorithm_state->object_dependencies, &obj_id,
+            sizeof(obj_id), entry);
+  if (entry == NULL) {
+    entry = malloc(sizeof(object_dependency));
+    entry->object_id = obj_id;
+    utarray_new(entry->dependent_tasks, &task_queue_entry_icd);
+    HASH_ADD(dependency_handle, algorithm_state->object_dependencies, object_id,
+             sizeof(entry->object_id), entry);
+  }
+  utarray_push_back(entry->dependent_tasks, &task_entry);
+}
+
+/**
+ * Record a queued task's missing object dependencies.  The given task should
+ * have at least one unsatisfied dependency.
+ *
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @param task_entry The task's queue entry.
+ * @returns Void.
+ */
+void add_task_dependencies(local_scheduler_state *state,
+                           scheduling_algorithm_state *algorithm_state,
+                           task_queue_entry *task_entry) {
+  task_spec *task = task_entry->spec;
+  int64_t num_args = task_num_args(task);
+  bool has_dependency = false;
```

Comment:
The name `has_dependency` doesn't make sense to me. Perhaps have an int `num_unsatisfied_dependencies`, increment it in the `if (entry == NULL)`, and CHECK(num_unsatisfied_dependencies > 0)`?